### PR TITLE
Change from openpolicyagent to quay.io/gateekeeper

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
Change the image from openpolicyagent to quay.io/gateekeeper For now, our dev is using a gatekeeper image from openpolicyagent/gatekeeper in gatekeeper-operator This should be changed to stolostron
Ref: https://issues.redhat.com/browse/ACM-9099
Signed-off-by: yiraeChristineKim <yikim@redhat.com>

### CODE REVIEW

**Comment by yiraeChristineKim** [comment]
@dhaiducek  I have thought about "LGTM! Do we also want to update to use github.com/stolostron/gatekeeper as well in this PR?"  I guess for now we don't need this because we haven't changed manifests yet

File: Makefile
Comment (dhaiducek) [inline Makefile:156]: Thinking more about this, this image update should be handled in `stolostron/gatekeeper` and then we just update the `IMPORT_MANIFESTS_PATH` here.

Comment (yiraeChristineKim) [reply Makefile:156]: so we need to direct IMPORT_MANIFESTS_PAT=https://github.com/open-policy-agent/gatekeeper  to  https://github.com/stolostron/gatekeeper after we change the https://github.com/stolostron/gatekeeper

Comment (dhaiducek) [reply Makefile:156]: Yes, that's what I'm thinking.

File: Makefile
Comment (dhaiducek) [inline Makefile:155]: Why was this changed? I think we still want to use the tag here rather than just the latest commit?

**Review by dhaiducek — COMMENTED** [review]
LGTM! I only have one question about cloning tags vs branches.

File: Makefile
Comment (yiraeChristineKim) [reply Makefile:155]: because I wanted to clone from `stolostron/gatekeeper` branch `release-3.X `

Comment (yiraeChristineKim) [reply Makefile:155]: Do you want tag?

Comment (yiraeChristineKim) [reply Makefile:155]: @dhaiducek Dale I changed to `release-$(shell echo $(VERSION)` again because CI failed when a tag does not include our own commits(stolostron commits)

Comment (dhaiducek) [reply Makefile]: Okay. And actually, thinking about it, we probably should be just adding a `-#` suffix to tags in `stolostron/gatekeeper` rather than deleting and recreating the same tag. I'm thinking let's merge it this way and then update the `build-push-stolostron` workflow to remove the trailing suffix so that it's overwriting the image tag each time. What do you think of that?

Comment (yiraeChristineKim) [reply Makefile]: Quay gatekeeper's tag going to be v3.X.X-X like this?

File: Makefile
Comment (mprahl) [inline Makefile:155]: Did you mean to change this?

File: Makefile
Comment (dhaiducek) [inline Makefile:157]:
```suggestion
		echo "The gatekeeper version $(GATEKEEPER_VERSION) is not in the list of tags at $(IMPORT_MANIFESTS_PATH)" ;\
```

File: Makefile
Comment (yiraeChristineKim) [reply Makefile]: yes!

**Comment by yiraeChristineKim** [comment]
@dhaiducek 
I tried 
```
[ "3.14.0" == "${GATEKEEPER_TAG%-*}" ] || { echo "error: Gatekeeper tag v3.14.0-3 doesn't align with Gatekeeper version 3.14.0 set in the operator."; exit 1; }
error: Gatekeeper tag v3.14.0-3 doesn't align with Gatekeeper version 3.14.0 set in the operator.
make: *** [tt] Error 1
```

seems like GATEKEEPER_TAG%-* doens't work properly

File: Makefile
Comment (dhaiducek) [inline Makefile:159]: Could we add a check that the GATEKEEPER_VERSION here matches the latest tag (in case we upgrade and forget to update it here):
```suggestion
	fi 
	if [ "$(GATEKEEPER_VERSION)" != "$${GATEKEEPER_TAG%-*}" ]; then \
	  echo "error: Gatekeeper tag $(GATEKEEPER_TAG) doesn't align with Gatekeeper version $(GATEKEEPER_VERSION) set in the operator.";
	  exit 1; \
	fi
```

File: Makefile
Comment (dhaiducek) [inline Makefile:168]: Extra marker?
```suggestion
	cd $(MAKEFILE_DIR)/$(GATEKEEPER_MANIFEST_DIR) && $(KUSTOMIZE) edit add resource *.yaml
```

**Review by dhaiducek — COMMENTED** [review]
Looking really good! I have a couple comments to consider.

**Comment by dhaiducek** [comment]
> seems like GATEKEEPER_TAG%-* doens't work properly

You need a double dollar sign `$$` to escape it and pass it out of Make to the shell. (I've added a comment to that effect.)

File: Makefile
Comment (dhaiducek) [inline Makefile:157]:
```suggestion
		echo "error: Gatekeeper version v$(GATEKEEPER_VERSION) set in the operator doesn't match Gatekeeper tag $(shell echo $(GATEKEEPER_TAG) | cut -d '-' -f 1) from $(IMPORT_MANIFESTS_PATH)." ;\
```

File: Makefile
Comment (mprahl) [inline Makefile:152]: @dhaiducek do you know if we'll run into GitHub API rate limiting with this? It seems we would use an API request every time we executed the Makefile.

Comment (mprahl) [reply Makefile:152]: Also, I really don't want an HTTP request to be made every time I use the Makefile.

Comment (dhaiducek) [reply Makefile:152]: I'm not sure the variable is expanded every time since it's not referenced except in `import-manifests` (though I'm never 100% sure with Make... 🤷‍♂️ ) . But to your point, we could probably just use `git tag` and clone/fetch the repo tags instead of a targeted clone. Looks like we might be able to do something like (untested) this?
```
git clone --no-checkout ...
git fetch --tags
git checkout $(shell git tag -l "$(shell $(GATEKEEPER_VERSION) | cut -d '.' -f 1-2)*" | sort --version-sort | tail -1)
```

Comment (yiraeChristineKim) [reply Makefile:152]: @mprahl Why calling HTTP request is bad?

Comment (mprahl) [reply Makefile:152]: @dhaiducek since the variable is not defined in a target, I think it would be expanded every time you call Make.

Comment (mprahl) [reply Makefile:152]: Actually, it does seem like if the variable is not referenced it doesn't call the shell command.

Comment (mprahl) [reply Makefile:152]: > @mprahl Why calling HTTP request is bad?

It's not bad if it's required but when it can be avoided, it reduces the chance of failure. For example, if I'm developing without an internet connection, the GitHub API is down, GitHub API rate limiting, or the network connection is flaky.

Since I misunderstood when the shell command actually gets run, I think it's not as bad as I thought since I think it's only called under `import-manifests`.

Comment (yiraeChristineKim) [reply Makefile:152]: @dhaiducek  good to go?

Comment (yiraeChristineKim) [reply Makefile:152]: Screenshot 2024-04-23 at 7 55 50 AM
These are our tag I thought our tags should be like these

Comment (dhaiducek) [reply Makefile:152]: Right, but if the version tag referenced in the operator is v3.15.1, for example, but we've updated Gatekeeper to have a v3.15.2, we want v3.15.2 to be picked up here by grepping for v3.15 and for `import-manifests` to then fail when it picks up the correct v3.15.2 tag.

**Review by mprahl — APPROVED** [review]
/hold in case Dale wants to take another look

File: Makefile
Comment (dhaiducek) [inline Makefile:152]: One more thing, sorry: I did some tests, and this needs to be a simply expanded variable, otherwise the HTTP request gets called 3 times during `import-manifests` (that's the `:=` assignment). I think we can also silence the `curl` command, since I don't find the output all that helpful since it's a quick turnaround:
```suggestion
GATEKEEPER_TAG := $(shell curl -sL https://api.github.com/repos/stolostron/gatekeeper/tags | jq -r '.[].name' | sort --version-sort | grep $(shell echo $(GATEKEEPER_VERSION) | cut -d '.' -f 1-2) | tail -1 )
```
ref: https://www.gnu.org/software/make/manual/html_node/Simple-Assignment.html

**Review by dhaiducek — CHANGES_REQUESTED** [review]

File: Makefile
Comment (yiraeChristineKim) [reply Makefile:152]: Great Great Thanks!!

**Review by dhaiducek — APPROVED** [review]
LGTM! Thanks!

**Review by mprahl — APPROVED** [review]
/unhold

**Comment by dhaiducek** [comment]
Sorry, I didn't notice this was targeted at a release branch.

**Pull request merged**
